core/shared/src/main/scala/zio/ZLayer.scala

```diff
@@ -110,6 +110,13 @@ final class ZLayer[-RIn, +E, +ROut <: Has[_]] private (
   def mapError[E1](f: E => E1): ZLayer[RIn, E1, ROut] =
     fold(ZLayer.fromFunctionManyM(f andThen ZIO.failNow), ZLayer.identity)
 
+  /**
+   * Returns a managed effect that, if evaluated, will return the lazily
+   * computed result of this layer.
+   */
+  def memoize: ZManaged[RIn, Nothing, ZLayer[Any, E, ROut]] =
```

Comment:
Do you think we could keep `RIn` inside the `ZLayer`?

Reply:
I think so. Do you think we would want to make the same change to the `memoize` methods in general? I was hoping to be able to implement this in terms of `ZManaged#memoize` so we could keep all the resource management logic there but we could have a custom implementation.

Reply:
It's just an intuition for `ZLayer`, because if we don't do that, you can no longer use `>>>` and `++` for the inner `ZLayer`, So the memoization would force you to some how supply the layer dependencies not through ZLayer but through the outer ZManaged.

Reply:
Yes. Okay, let me work on this.

Reply:
@jdegoes I refactored `ZIO#memoize` and `ZManaged#memoize` to require the environment for the inner effect instead of the outer effect. I think this creates a nice consistency across these methods, `ZLayer#memoize`, and `ZIO#once`. It is also more general than the previous signature because you can always access the environment and provide it to the effect you are memoizing to get the old signature, but you can't go the other way around. Are you good with this?

@iravid If you have a chance could you take a look at the `ZManaged#memoize` implementation and of course if you think this makes sense?